File: src/main/tutorial.ts
import { BrowserWindow, screen } from 'electron';
import Store from 'electron-store';
import type { StoreSchema } from './store';

export type TutorialStep = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;

interface TutorialStepConfig {
  step: TutorialStep;
  copy: string | ((hotkeys: { openChat: string; openAssistant: string }) => string);
  fallbackDelay: number; // ms before showing hint/fallback
  autoAdvance?: boolean; // Step advances automatically (no user action needed)
  autoAdvanceDelay?: number; // ms to wait before auto-advancing
  delayBefore?: number; // ms to wait before showing this step
}

/**
 * Format hotkey for display (e.g., "CommandOrControl+Shift+Space" -> "Cmd+Shift+Space")
 */
function formatHotkey(hotkey: string): string {
  return hotkey
    .replace('CommandOrControl', process.platform === 'darwin' ? 'Cmd' : 'Ctrl')
    .replace('Command', 'Cmd')
    .replace('Control', 'Ctrl');
}

const TUTORIAL_STEPS: TutorialStepConfig[] = [
  { step: 1, copy: "Hi there! I'm Clawster, your desktop companion!", fallbackDelay: 0, autoAdvance: true, autoAdvanceDelay: 2500 },
  { step: 2, copy: "Try clicking on me to see how I react!", fallbackDelay: 3000, delayBefore: 500 },
  { step: 3, copy: "Fun right? I have lots of different reactions!", fallbackDelay: 0, autoAdvance: true, autoAdvanceDelay: 2500, delayBefore: 500 },
  { step: 4, copy: "Now move your mouse away from me...", fallbackDelay: 6000, delayBefore: 500 },
  { step: 5, copy: "I followed you! Don't worry - I only do this when I feel lonely. You can disable this in Settings.", fallbackDelay: 0, delayBefore: 500 },
  { step: 6, copy: "Want to chat? I'm always here to help!", fallbackDelay: 0, autoAdvance: true, autoAdvanceDelay: 2000, delayBefore: 500 },
  { step: 7, copy: (hotkeys) => `Press ${formatHotkey(hotkeys.openChat)} to chat with me anytime.`, fallbackDelay: 6000, delayBefore: 500 },
  { step: 8, copy: "You can also access Settings and chat history.", fallbackDelay: 0, autoAdvance: true, autoAdvanceDelay: 3500, delayBefore: 500 },
  { step: 9, copy: (hotkeys) => `Press ${formatHotkey(hotkeys.openAssistant)} to open the panel.`, fallbackDelay: 5000, delayBefore: 500 },
  { step: 10, copy: "That's it! I'll be right here if you need me.", fallbackDelay: 0, autoAdvance: true, autoAdvanceDelay: 3000, delayBefore: 500 },
];

export class TutorialManager {
  private store: Store<StoreSchema>;
  private petWindow: BrowserWindow | null = null;
  private currentStep: TutorialStep | null = null;
  private isActive = false;
  private cursorTrackingInterval: NodeJS.Timeout | null = null;
  private followInterval: NodeJS.Timeout | null = null;
  private fallbackTimer: NodeJS.Timeout | null = null;
  private animateMoveTo: ((x: number, y: number, duration: number) => Promise<void>) | null = null;
  private expandWindow: (() => void) | null = null;
  private contractWindow: (() => void) | null = null;

  constructor(store: Store<StoreSchema>) {
    this.store = store;
  }

  /**
   * Set the pet window reference
   */
  setPetWindow(window: BrowserWindow | null): void {
    this.petWindow = window;
  }

  /**
   * Set the animation function for moving the pet
   */
  setAnimateMoveTo(fn: (x: number, y: number, duration: number) => Promise<void>): void {
    this.animateMoveTo = fn;
  }

  /**
   * Set the window resize functions
   */
  setWindowResizeFunctions(expand: () => void, contract: () => void): void {
    this.expandWindow = expand;
    this.contractWindow = contract;
  }

  /**
   * Check if tutorial should start or show resume prompt
   */
  shouldStartTutorial(): boolean {
    const tutorial = this.store.get('tutorial');
    return !tutorial.completedAt;
  }

  /**
   * Check if tutorial was interrupted and should show resume prompt
   */
  shouldShowResumePrompt(): boolean {
    const tutorial = this.store.get('tutorial');
    return tutorial.wasInterrupted && !tutorial.completedAt;
  }

  /**
   * Start the tutorial from the beginning or resume
   */
  async start(fromStep: TutorialStep = 1): Promise<void> {
    if (!this.petWindow) {
      console.error('[Tutorial] Cannot start: petWindow not set');
      return;
    }

    this.isActive = true;
    this.store.set('tutorial.wasInterrupted', true);
    this.store.set('tutorial.lastStep', fromStep);

    // Move pet to center of screen first
    await this.movePetToCenter();

    // Expand window for tutorial UI
    this.expandWindow?.();

    this.goToStep(fromStep);
    console.log(`[Tutorial] Started at step ${fromStep}`);
  }

  /**
   * Move pet to center of screen
   */
  private async movePetToCenter(): Promise<void> {
    if (!this.petWindow || !this.animateMoveTo) return;

    const { width: screenWidth, height: screenHeight } = screen.getPrimaryDisplay().workAreaSize;
    const [petWidth, petHeight] = this.petWindow.getSize();

    const centerX = Math.round((screenWidth - petWidth) / 2);
    const centerY = Math.round((screenHeight - petHeight) / 2);

    await this.animateMoveTo(centerX, centerY, 1000);
  }

  /**
   * Move pet back to bottom right corner
   */
  private async movePetToBottomRight(): Promise<void> {
    if (!this.petWindow || !this.animateMoveTo) return;

    const { width: screenWidth, height: screenHeight } = screen.getPrimaryDisplay().workAreaSize;
    const petWidth = 130; // Normal pet window size
    const petHeight = 130;

    const targetX = screenWidth - petWidth - 20;
    const targetY = screenHeight - petHeight - 20;

    await this.animateMoveTo(targetX, targetY, 1000);
  }

  /**
   * Get current tutorial state
   */
  getStatus(): { isActive: boolean; currentStep: TutorialStep | null; completed: boolean } {
    const tutorial = this.store.get('tutorial');
    return {
      isActive: this.isActive,
      currentStep: this.currentStep,
      completed: !!tutorial.completedAt,
    };
  }

  /**
   * Handle pet clicked (for step 2)
   */
  handlePetClicked(): void {
    if (!this.isActive || this.currentStep !== 2) return;
    this.advanceStep();
  }

  /**
   * Handle "Next" button clicked (for all steps)
   */
  handleNextClicked(): void {
    if (!this.isActive || !this.currentStep) return;
    this.advanceStep();
  }

  /**
   * Handle hotkey pressed - check if it matches current step requirement
   */
  handleHotkeyPressed(hotkey: 'openChat' | 'openAssistant'): void {
    if (!this.isActive) return;

    if (this.currentStep === 7 && hotkey === 'openChat') {
      this.advanceStep();
    } else if (this.currentStep === 9 && hotkey === 'openAssistant') {
      this.advanceStep();
    }
  }

  /**
   * Handle "Open Panel" button clicked (for step 9 fallback)
   */
  handleOpenPanelClicked(): void {
    if (!this.isActive || this.currentStep !== 9) return;
    this.advanceStep();
  }

  /**
   * Skip the tutorial
   */
  async skip(): Promise<void> {
    this.cleanup();
    this.isActive = false;
    this.currentStep = null;
    this.store.set('tutorial.completedAt', new Date().toISOString());
    this.store.set('tutorial.wasInterrupted', false);

    // Contract window back to normal first
    this.contractWindow?.();

    // Move pet back to bottom right corner
    await this.movePetToBottomRight();

    this.petWindow?.webContents.send('tutorial-ended', { skipped: true });
    console.log('[Tutorial] Skipped');
  }

  /**
   * Resume tutorial after app restart
   */
  resume(): void {
    const tutorial = this.store.get('tutorial');
    // Start from beginning for simplicity (avoids weird partial states)
    this.start(1);
    this.petWindow?.webContents.send('tutorial-resumed');
  }

  /**
   * Start over from beginning
   */
  startOver(): void {
    this.store.set('tutorial.lastStep', 0);
    this.store.set('tutorial.wasInterrupted', false);
    this.start(1);
  }

  /**
   * Replay tutorial (called from settings)
   */
  replay(): void {
    this.store.set('tutorial.completedAt', null);
    this.store.set('tutorial.lastStep', 0);
    this.store.set('tutorial.wasInterrupted', false);
    this.start(1);
    console.log('[Tutorial] Replaying');
  }

  /**
   * Get the hotkeys from the store
   */
  private getHotkeys(): { openChat: string; openAssistant: string } {
    return {
      openChat: this.store.get('hotkeys.openChat') as string,
      openAssistant: this.store.get('hotkeys.openAssistant') as string,
    };
  }

  /**
   * Resolve step copy (handles both static strings and dynamic functions)
   */
  private resolveCopy(copy: string | ((hotkeys: { openChat: string; openAssistant: string }) => string)): string {
    if (typeof copy === 'function') {
      return copy(this.getHotkeys());
    }
    return copy;
  }

  /**
   * Go to a specific step
   */
  private goToStep(step: TutorialStep): void {
    this.cleanup();

    const stepConfig = TUTORIAL_STEPS.find(s => s.step === step);
    if (!stepConfig) {
      console.error(`[Tutorial] Unknown step: ${step}`);
      return;
    }

    // Apply delay before showing step
    const delayBefore = stepConfig.delayBefore || 0;

    setTimeout(() => {
      this.currentStep = step;
      this.store.set('tutorial.lastStep', step);

      // Resolve the copy (handles dynamic hotkeys)
      const resolvedCopy = this.resolveCopy(stepConfig.copy);

      // Notify renderer of new step
      this.petWindow?.webContents.send('tutorial-step', {
        step,
        copy: resolvedCopy,
        totalSteps: TUTORIAL_STEPS.length,
      });

      // Set up step-specific logic
      this.setupStepLogic(step, stepConfig);
      console.log(`[Tutorial] Now at step ${step}: ${resolvedCopy}`);
    }, delayBefore);
  }

  /**
   * Set up the logic for a specific step
   */
  private setupStepLogic(step: TutorialStep, config: TutorialStepConfig): void {
    // Set up fallback timer if needed
    if (config.fallbackDelay > 0) {
      this.fallbackTimer = setTimeout(() => {
        this.showHint(step);
      }, config.fallbackDelay);
    }

    // Handle auto-advance with configurable delay
    if (config.autoAdvance && config.autoAdvanceDelay) {
      setTimeout(() => {
        if (this.currentStep === step) {
          this.advanceStep();
        }
      }, config.autoAdvanceDelay);
    }

    // Step-specific setup
    switch (step) {
      case 4:
        // Move away step - start cursor tracking
        this.startCursorTracking();
        break;
      case 10:
        // Final step - wave animation
        this.performFinalAnimation();
        break;
    }
  }

  /**
   * Show hint for current step
   */
  private showHint(step: TutorialStep): void {
    let hintType: string;

    switch (step) {
      case 2:
        // Click on me step
        hintType = 'pulse';
        break;
      case 4:
        // Move away step
        hintType = 'arrow';
        break;
      case 7:
        // Chat hotkey step
        hintType = 'skip-button';
        break;
      case 9:
        // Panel hotkey step
        hintType = 'open-panel-button';
        break;
      default:
        return;
    }

    this.petWindow?.webContents.send('tutorial-hint', { step, hintType });
    console.log(`[Tutorial] Showing hint for step ${step}: ${hintType}`);
  }

  /**
   * Advance to next step
   */
  private advanceStep(): void {
    if (!this.currentStep) return;

    const nextStep = (this.currentStep + 1) as TutorialStep;

    if (nextStep > 10) {
      this.complete();
    } else {
      this.goToStep(nextStep);
    }
  }

  /**
   * Complete the tutorial
   */
  private async complete(): Promise<void> {
    this.cleanup();
    this.isActive = false;
    this.currentStep = null;
    this.store.set('tutorial.completedAt', new Date().toISOString());
    this.store.set('tutorial.wasInterrupted', false);

    // Contract window back to normal first
    this.contractWindow?.();

    // Move pet back to bottom right corner
    await this.movePetToBottomRight();

    this.petWindow?.webContents.send('tutorial-ended', { skipped: false });
    console.log('[Tutorial] Completed!');
  }

  /**
   * Start tracking cursor distance for step 4
   */
  private startCursorTracking(): void {
    this.cursorTrackingInterval = setInterval(() => {
      if (!this.petWindow || this.currentStep !== 4) return;

      const cursor = screen.getCursorScreenPoint();
      const [petX, petY] = this.petWindow.getPosition();
      const [petWidth, petHeight] = this.petWindow.getSize();

      // Calculate distance from cursor to pet center
      const petCenterX = petX + petWidth / 2;
      const petCenterY = petY + petHeight / 2;
      const distance = Math.sqrt(
        Math.pow(cursor.x - petCenterX, 2) + Math.pow(cursor.y - petCenterY, 2)
      );

      // If cursor is far enough, start following behavior
      if (distance > 200 && !this.followInterval) {
        this.startFollowBehavior();
      }
    }, 100);
  }

  /**
   * Start follow behavior for step 4→5 transition
   * Similar to attention seeker - slower, playful movement
   */
  private startFollowBehavior(): void {
    if (this.followInterval || !this.animateMoveTo) return;

    let followCount = 0;
    const maxFollows = 2;
    let isMoving = false;

    // Set excited mood when starting to follow
    this.petWindow?.webContents.send('clawbot-mood', { state: 'excited', reason: 'following cursor' });

    // Do first move after a short delay
    const doFollow = async () => {
      if (!this.petWindow || !this.animateMoveTo || this.currentStep !== 4 || isMoving) {
        return;
      }

      const cursor = screen.getCursorScreenPoint();
      const [petX, petY] = this.petWindow.getPosition();
      const { width: screenWidth, height: screenHeight } = screen.getPrimaryDisplay().workAreaSize;

      // Calculate current distance from cursor to pet
      const distance = Math.sqrt(
        Math.pow(cursor.x - petX, 2) + Math.pow(cursor.y - petY, 2)
      );

      // Only move if cursor is far enough away (similar to attention seeker's 600px threshold)
      if (distance < 300) {
        return;
      }

      isMoving = true;

      // Calculate position near cursor (with offset so pet doesn't cover cursor)
      // Same as attention seeker: offset of 80px
      const offset = 80;
      let targetX = cursor.x + offset;
      let targetY = cursor.y + offset;

      // Keep within screen bounds (using 300 like attention seeker)
      targetX = Math.max(0, Math.min(targetX, screenWidth - 300));
      targetY = Math.max(0, Math.min(targetY, screenHeight - 300));

      console.log(`[Tutorial] Following cursor to (${targetX}, ${targetY}), distance was ${Math.round(distance)}px`);

      // Animate move - same 1500ms as attention seeker
      await this.animateMoveTo(targetX, targetY, 1500);

      isMoving = false;
      followCount++;

      // Check current distance after move
      const newCursor = screen.getCursorScreenPoint();
      const newDistance = Math.sqrt(
        Math.pow(newCursor.x - targetX, 2) +
        Math.pow(newCursor.y - targetY, 2)
      );

      console.log(`[Tutorial] After move, distance is ${Math.round(newDistance)}px, followCount=${followCount}`);

      // If we've "caught" the cursor or done enough follows, advance
      if (newDistance < 150 || followCount >= maxFollows) {
        this.stopFollowBehavior();
        // Move back to center before showing step 3
        await this.movePetToCenter();
        // Brief pause before advancing
        setTimeout(() => {
          this.advanceStep();
        }, 300);
      }
    };

    // Use similar timing to attention seeker - check every 3-5 seconds
    this.followInterval = setInterval(doFollow, 4000);

    // Do first check after 1 second
    setTimeout(doFollow, 1000);
  }

  /**
   * Stop follow behavior
   */
  private stopFollowBehavior(): void {
    if (this.followInterval) {
      clearInterval(this.followInterval);
      this.followInterval = null;
    }
  }

  /**
   * Perform final animation for step 10
   */
  private performFinalAnimation(): void {
    // Send wave mood to pet
    this.petWindow?.webContents.send('clawbot-mood', { state: 'happy' });
  }

  /**
   * Clean up timers and intervals
   */
  private cleanup(): void {
    if (this.cursorTrackingInterval) {
      clearInterval(this.cursorTrackingInterval);
      this.cursorTrackingInterval = null;
    }
    if (this.followInterval) {
      clearInterval(this.followInterval);
      this.followInterval = null;
    }
    if (this.fallbackTimer) {
      clearTimeout(this.fallbackTimer);
      this.fallbackTimer = null;
    }
  }

  /**
   * Clean up when app quits
   */
  destroy(): void {
    this.cleanup();
    this.isActive = false;
    this.currentStep = null;
  }
}

File: src-tauri/src/tutorial.rs
use serde::Serialize;
use std::sync::atomic::{AtomicBool, AtomicU8, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tauri::{AppHandle, Emitter, Manager, Wry};

use crate::pet;
use crate::store::AppStore;
use crate::windows;

// Tutorial step config matching Electron
struct StepConfig {
    step: u8,
    copy: StepCopy,
    fallback_delay_ms: u64,
    auto_advance: bool,
    auto_advance_delay_ms: u64,
    delay_before_ms: u64,
}

enum StepCopy {
    Static(&'static str),
    Dynamic(fn(&str, &str) -> String),
}

fn format_hotkey(hotkey: &str) -> String {
    hotkey
        .replace("CommandOrControl", if cfg!(target_os = "macos") { "Cmd" } else { "Ctrl" })
        .replace("Command", "Cmd")
        .replace("Control", "Ctrl")
}

fn step7_copy(open_chat: &str, _open_assistant: &str) -> String {
    format!("Press {} to chat with me anytime.", format_hotkey(open_chat))
}

fn step9_copy(_open_chat: &str, open_assistant: &str) -> String {
    format!("Press {} to open the panel.", format_hotkey(open_assistant))
}

const TUTORIAL_STEPS: &[StepConfig] = &[
    StepConfig { step: 1, copy: StepCopy::Static("Hi there! I'm Clawster, your desktop companion!"), fallback_delay_ms: 0, auto_advance: true, auto_advance_delay_ms: 2500, delay_before_ms: 0 },
    StepConfig { step: 2, copy: StepCopy::Static("Try clicking on me to see how I react!"), fallback_delay_ms: 3000, auto_advance: false, auto_advance_delay_ms: 0, delay_before_ms: 500 },
    StepConfig { step: 3, copy: StepCopy::Static("Fun right? I have lots of different reactions!"), fallback_delay_ms: 0, auto_advance: true, auto_advance_delay_ms: 2500, delay_before_ms: 500 },
    StepConfig { step: 4, copy: StepCopy::Static("Now move your mouse away from me..."), fallback_delay_ms: 6000, auto_advance: false, auto_advance_delay_ms: 0, delay_before_ms: 500 },
    StepConfig { step: 5, copy: StepCopy::Static("I followed you! Don't worry - I only do this when I feel lonely. You can disable this in Settings."), fallback_delay_ms: 0, auto_advance: false, auto_advance_delay_ms: 0, delay_before_ms: 500 },
    StepConfig { step: 6, copy: StepCopy::Static("Want to chat? I'm always here to help!"), fallback_delay_ms: 0, auto_advance: true, auto_advance_delay_ms: 2000, delay_before_ms: 500 },
    StepConfig { step: 7, copy: StepCopy::Dynamic(step7_copy), fallback_delay_ms: 6000, auto_advance: false, auto_advance_delay_ms: 0, delay_before_ms: 500 },
    StepConfig { step: 8, copy: StepCopy::Static("You can also access Settings and chat history."), fallback_delay_ms: 0, auto_advance: true, auto_advance_delay_ms: 3500, delay_before_ms: 500 },
    StepConfig { step: 9, copy: StepCopy::Dynamic(step9_copy), fallback_delay_ms: 5000, auto_advance: false, auto_advance_delay_ms: 0, delay_before_ms: 500 },
    StepConfig { step: 10, copy: StepCopy::Static("That's it! I'll be right here if you need me."), fallback_delay_ms: 0, auto_advance: true, auto_advance_delay_ms: 3000, delay_before_ms: 500 },
];

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct TutorialStepEvent {
    step: u8,
    copy: String,
    total_steps: usize,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct TutorialHintEvent {
    step: u8,
    hint_type: String,
}

#[derive(Debug, Clone, Serialize)]
struct TutorialEndedEvent {
    skipped: bool,
}

pub struct TutorialManager {
    is_active: Arc<AtomicBool>,
    current_step: Arc<AtomicU8>,
}

impl TutorialManager {
    pub fn new() -> Self {
        Self {
            is_active: Arc::new(AtomicBool::new(false)),
            current_step: Arc::new(AtomicU8::new(0)),
        }
    }

    pub fn is_active(&self) -> bool {
        self.is_active.load(Ordering::Relaxed)
    }

    pub fn current_step(&self) -> u8 {
        self.current_step.load(Ordering::Relaxed)
    }

    pub fn should_start_tutorial(app: &AppHandle<Wry>) -> bool {
        let store = app.state::<AppStore>();
        let completed_at = store.get_value("tutorial.completedAt");
        completed_at.map_or(true, |v| v.is_null())
    }

    pub fn should_show_resume_prompt(app: &AppHandle<Wry>) -> bool {
        let store = app.state::<AppStore>();
        let interrupted = store.get_value("tutorial.wasInterrupted")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        let completed = store.get_value("tutorial.completedAt")
            .map_or(false, |v| !v.is_null());
        interrupted && !completed
    }

    pub fn start(&self, app: &AppHandle<Wry>, from_step: u8) {
        self.is_active.store(true, Ordering::Relaxed);
        let store = app.state::<AppStore>();
        let _ = store.set_value("tutorial.wasInterrupted", serde_json::json!(true));
        let _ = store.set_value("tutorial.lastStep", serde_json::json!(from_step));

        // Expand pet window for tutorial
        expand_pet_window(app);

        self.go_to_step(app, from_step);
    }

    pub fn skip(&self, app: &AppHandle<Wry>) {
        self.is_active.store(false, Ordering::Relaxed);
        self.current_step.store(0, Ordering::Relaxed);

        let store = app.state::<AppStore>();
        let _ = store.set_value("tutorial.completedAt", serde_json::json!(chrono::Utc::now().to_rfc3339()));
        let _ = store.set_value("tutorial.wasInterrupted", serde_json::json!(false));

        contract_pet_window(app);
        let _ = app.emit("tutorial-ended", TutorialEndedEvent { skipped: true });
    }

    pub fn handle_pet_clicked(&self, app: &AppHandle<Wry>) {
        if !self.is_active() || self.current_step() != 2 {
            return;
        }
        self.advance_step(app);
    }

    pub fn handle_next(&self, app: &AppHandle<Wry>) {
        if !self.is_active() || self.current_step() == 0 {
            return;
        }
        self.advance_step(app);
    }

    pub fn handle_hotkey_pressed(&self, app: &AppHandle<Wry>, hotkey: &str) {
        if !self.is_active() {
            return;
        }
        if self.current_step() == 7 && hotkey == "openChat" {
            self.advance_step(app);
        } else if self.current_step() == 9 && hotkey == "openAssistant" {
            self.advance_step(app);
        }
    }

    pub fn handle_open_panel(&self, app: &AppHandle<Wry>) {
        if !self.is_active() || self.current_step() != 9 {
            return;
        }
        self.advance_step(app);
    }

    pub fn replay(&self, app: &AppHandle<Wry>) {
        let store = app.state::<AppStore>();
        let _ = store.set_value("tutorial.completedAt", serde_json::json!(null));
        let _ = store.set_value("tutorial.lastStep", serde_json::json!(0));
        let _ = store.set_value("tutorial.wasInterrupted", serde_json::json!(false));
        self.start(app, 1);
    }

    pub fn get_status(&self, app: &AppHandle<Wry>) -> serde_json::Value {
        let store = app.state::<AppStore>();
        let completed = store.get_value("tutorial.completedAt")
            .map_or(false, |v| !v.is_null());
        let current = if self.is_active() { Some(self.current_step()) } else { None };
        serde_json::json!({
            "isActive": self.is_active(),
            "currentStep": current,
            "completed": completed
        })
    }

    fn go_to_step(&self, app: &AppHandle<Wry>, step: u8) {
        let config = match TUTORIAL_STEPS.get((step - 1) as usize) {
            Some(c) => c,
            None => return,
        };

        let delay_before = config.delay_before_ms;
        let app2 = app.clone();
        let step_val = step;

        // Get hotkeys for dynamic copy
        let store = app.state::<AppStore>();
        let open_chat = store.get_value("hotkeys.openChat")
            .and_then(|v| v.as_str().map(String::from))
            .unwrap_or_else(|| "CommandOrControl+Shift+Space".to_string());
        let open_assistant = store.get_value("hotkeys.openAssistant")
            .and_then(|v| v.as_str().map(String::from))
            .unwrap_or_else(|| "CommandOrControl+Shift+A".to_string());

        let copy_text = match &config.copy {
            StepCopy::Static(s) => s.to_string(),
            StepCopy::Dynamic(f) => f(&open_chat, &open_assistant),
        };

        let auto_advance = config.auto_advance;
        let auto_delay = config.auto_advance_delay_ms;
        let fallback_delay = config.fallback_delay_ms;

        let is_active = self.is_active.clone();
        let current_step = self.current_step.clone();

        tauri::async_runtime::spawn(async move {
            if delay_before > 0 {
                tokio::time::sleep(Duration::from_millis(delay_before)).await;
            }

            if !is_active.load(Ordering::Relaxed) {
                return;
            }

            current_step.store(step_val, Ordering::Relaxed);
            let _ = app2.state::<AppStore>().set_value("tutorial.lastStep", serde_json::json!(step_val));

            let _ = app2.emit("tutorial-step", TutorialStepEvent {
                step: step_val,
                copy: copy_text,
                total_steps: TUTORIAL_STEPS.len(),
            });

            // Auto-advance timer
            if auto_advance && auto_delay > 0 {
                let app3 = app2.clone();
                let is_active2 = is_active.clone();
                let cs2 = current_step.clone();
                tauri::async_runtime::spawn(async move {
                    tokio::time::sleep(Duration::from_millis(auto_delay)).await;
                    if is_active2.load(Ordering::Relaxed) && cs2.load(Ordering::Relaxed) == step_val {
                        advance_step_async(&app3, &is_active2, &cs2).await;
                    }
                });
            }

            // Fallback hint timer
            if fallback_delay > 0 {
                let app4 = app2.clone();
                let is_active3 = is_active.clone();
                let cs3 = current_step.clone();
                tauri::async_runtime::spawn(async move {
                    tokio::time::sleep(Duration::from_millis(fallback_delay)).await;
                    if is_active3.load(Ordering::Relaxed) && cs3.load(Ordering::Relaxed) == step_val {
                        let hint_type = match step_val {
                            2 => "pulse",
                            4 => "arrow",
                            7 => "skip-button",
                            9 => "open-panel-button",
                            _ => return,
                        };
                        let _ = app4.emit("tutorial-hint", TutorialHintEvent {
                            step: step_val,
                            hint_type: hint_type.to_string(),
                        });
                    }
                });
            }
        });
    }

    fn advance_step(&self, app: &AppHandle<Wry>) {
        let next = self.current_step() + 1;
        if next > 10 {
            self.complete(app);
        } else {
            self.go_to_step(app, next);
        }
    }

    fn complete(&self, app: &AppHandle<Wry>) {
        self.is_active.store(false, Ordering::Relaxed);
        self.current_step.store(0, Ordering::Relaxed);

        let store = app.state::<AppStore>();
        let _ = store.set_value("tutorial.completedAt", serde_json::json!(chrono::Utc::now().to_rfc3339()));
        let _ = store.set_value("tutorial.wasInterrupted", serde_json::json!(false));

        contract_pet_window(app);
        let _ = app.emit("tutorial-ended", TutorialEndedEvent { skipped: false });
    }
}

async fn advance_step_async(
    app: &AppHandle<Wry>,
    is_active: &AtomicBool,
    current_step: &AtomicU8,
) {
    let next = current_step.load(Ordering::Relaxed) + 1;
    if next > 10 {
        // Complete
        is_active.store(false, Ordering::Relaxed);
        current_step.store(0, Ordering::Relaxed);
        let store = app.state::<AppStore>();
        let _ = store.set_value("tutorial.completedAt", serde_json::json!(chrono::Utc::now().to_rfc3339()));
        let _ = store.set_value("tutorial.wasInterrupted", serde_json::json!(false));
        contract_pet_window(app);
        let _ = app.emit("tutorial-ended", TutorialEndedEvent { skipped: false });
    } else {
        // Go to next step (simplified inline version)
        let config = match TUTORIAL_STEPS.get((next - 1) as usize) {
            Some(c) => c,
            None => return,
        };

        if config.delay_before_ms > 0 {
            tokio::time::sleep(Duration::from_millis(config.delay_before_ms)).await;
        }

        current_step.store(next, Ordering::Relaxed);

        let store = app.state::<AppStore>();
        let open_chat = store.get_value("hotkeys.openChat")
            .and_then(|v| v.as_str().map(String::from))
            .unwrap_or_else(|| "CommandOrControl+Shift+Space".to_string());
        let open_assistant = store.get_value("hotkeys.openAssistant")
            .and_then(|v| v.as_str().map(String::from))
            .unwrap_or_else(|| "CommandOrControl+Shift+A".to_string());

        let copy_text = match &config.copy {
            StepCopy::Static(s) => s.to_string(),
            StepCopy::Dynamic(f) => f(&open_chat, &open_assistant),
        };

        let _ = app.emit("tutorial-step", TutorialStepEvent {
            step: next,
            copy: copy_text,
            total_steps: TUTORIAL_STEPS.len(),
        });
    }
}

// Window expand/contract for tutorial UI

fn expand_pet_window(app: &AppHandle<Wry>) {
    let Some(pet) = app.get_webview_window(windows::LABEL_PET) else { return };
    let pos = pet.outer_position().unwrap_or_default();
    let size = pet.outer_size().unwrap_or_default();
    let scale = pet.scale_factor().unwrap_or(1.0);

    let current_w = size.width as f64 / scale;
    let current_h = size.height as f64 / scale;
    let current_x = pos.x as f64 / scale;
    let current_y = pos.y as f64 / scale;

    let new_y = current_y - (windows::PET_WINDOW_TUTORIAL_HEIGHT - current_h);
    let new_x = current_x - (windows::PET_WINDOW_TUTORIAL_WIDTH - current_w) / 2.0;

    let _ = pet.set_size(tauri::LogicalSize::new(
        windows::PET_WINDOW_TUTORIAL_WIDTH,
        windows::PET_WINDOW_TUTORIAL_HEIGHT,
    ));
    let _ = pet.set_position(tauri::LogicalPosition::new(
        f64::max(0.0, new_x.round()),
        f64::max(0.0, new_y.round()),
    ));
    let _ = app.emit("tutorial-window-expanded", true);
}

fn contract_pet_window(app: &AppHandle<Wry>) {
    let Some(pet) = app.get_webview_window(windows::LABEL_PET) else { return };
    let pos = pet.outer_position().unwrap_or_default();
    let scale = pet.scale_factor().unwrap_or(1.0);

    let current_x = pos.x as f64 / scale;
    let current_y = pos.y as f64 / scale;

    let new_y = current_y + (windows::PET_WINDOW_TUTORIAL_HEIGHT - windows::PET_WINDOW_HEIGHT);
    let new_x = current_x + (windows::PET_WINDOW_TUTORIAL_WIDTH - windows::PET_WINDOW_WIDTH) / 2.0;

    let _ = pet.set_size(tauri::LogicalSize::new(
        windows::PET_WINDOW_WIDTH,
        windows::PET_WINDOW_HEIGHT,
    ));
    let _ = pet.set_position(tauri::LogicalPosition::new(new_x.round(), new_y.round()));
    let _ = app.emit("tutorial-window-expanded", false);
}
